kratos/geometries/nurbs_curve_shape_functions.h

```diff
@@ -0,0 +1,263 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Thomas Oberbichler
+//
+//  Ported from the ANurbs library (https://github.com/oberbichler/ANurbs)
+//
+
+#if !defined(KRATOS_NURBS_CURVE_SHAPE_FUNCTIONS_H_INCLUDED )
+#define  KRATOS_NURBS_CURVE_SHAPE_FUNCTIONS_H_INCLUDED
+
+#include "nurbs_utility.h"
+
+#include <vector>
+
+namespace Kratos {
+
+class NurbsCurveShapeFunction
+{
+private:    // variables
+    int mDegree;
+    int mOrder;
+    std::vector<double> mValues;
+    std::vector<double> mLeft;
+    std::vector<double> mRight;
+    std::vector<double> mNdu;
+    std::vector<double> mA;
+    std::vector<double> mB;
+    int mFirstNonzeroPole;
+
+private:    // methods
+    double& Value(const int Order, const int Pole)
+    {
+        const int index = NurbsUtility::GetSingleIndex(GetNbShapeFunctions(),
+            GetNbNonzeroPoles(), Order, Pole);
+
+        return mValues[index];
+    }
+
+    double& Ndu(const int IndexI, const int IndexJ)
+    {
+        const int index = NurbsUtility::GetSingleIndex(GetNbShapeFunctions(),
+            GetNbNonzeroPoles(), IndexI, IndexJ);
+
+        return mNdu[index];
+    }
+
+    void ClearValues()
+    {
+        const int nb_values = GetNbNonzeroPoles() * GetNbShapeFunctions();
+
+        std::fill(mValues.begin(), mValues.begin() + nb_values, 0);
+    }
+
+public:     // constructors
+    NurbsCurveShapeFunction()
+    {
+    }
+
+    NurbsCurveShapeFunction(const int Degree, const int Order)
+    {
+        Resize(Degree, Order);
+    }
+
+public:     // methods
+    void Resize(const int Degree, const int Order)
+    {
+        mValues.resize((Order + 1) * (Degree + 1));
+        mLeft.resize(Degree);
+        mRight.resize(Degree);
+        mNdu.resize((Degree + 1) * (Degree + 1));
+        mA.resize(Degree + 1);
+        mB.resize(Degree + 1);
+
+        mDegree = Degree;
+        mOrder = Order;
+    }
+
+    int GetDegree() const
+    {
+        return mDegree;
+    }
+
+    int GetOrder() const
+    {
+        return mOrder;
+    }
+
+    int GetNbNonzeroPoles() const
+    {
+        return GetDegree() + 1;
+    }
+
+    int GetNbShapeFunctions() const
+    {
+        return GetOrder() + 1;
+    }
+
+    double operator()(const int Order, const int Pole) const
+    {
+        return Value(Order, Pole);
+    }
+
+    double Value(const int Order, const int Pole) const
```

Comment:
Here it is not the pole but the control point, because here the weight is included, isn't it?

Reply:
Maybe I should name it `PoleIndex` to make it clear for anyone

Reply:
We can also follow the Kratos style of 
typedef std::size_t IndexType;